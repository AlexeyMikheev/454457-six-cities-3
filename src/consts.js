
export const Feature = {
  WIFI: `Wi-Fi`,
  WASHINGMACHINE: `Washing machine`,
  TOWELS: `Towels`,
  HEATING: `Heating`,
  COFFEEMACHINE: `Coffee machine`,
  BABYSET: `Baby seat`,
  KITCHEN: `Kitchen`,
  DISHWASHER: `Dishwasher`,
  CABELTV: `Cabel TV`,
  FRIDGE: `Fridge`
};

export const FEATURES = [
  Feature.WIFI,
  Feature.WASHINGMACHINE,
  Feature.HEATING,
  Feature.COFFEEMACHINE,
  Feature.BABYSET,
  Feature.DISHWASHER,
  Feature.CABELTV,
  Feature.FRIDGE
];

export const OfferType = {
  APARTMENT: `Apartment`,
  ROOM: `Private room`,
  HOUSE: `House`,
  HOTEL: `Hotel`
};

export const OFFERTYPES = [
  OfferType.APARTMENT,
  OfferType.ROOM,
  OfferType.HOUSE,
  OfferType.HOTEL
];

export const ViewMode = {
  Main: 0,
  Property: 1
};

export const VIEWMODES = [
  ViewMode.Main,
  ViewMode.Property
];

export const MIN_RATING = 1;
export const MAX_RATING = 5;

export const MAX_REVIEWS_DISPLAY_COUNT = 10;

export const MAX_IMAGES_DISPLAY_COUNT = 6;
export const MAX_NEAR_DISPLAY_COUNT = 3;

export const SortType = {
  POPULAR: `Popular`,
  PRICE_LH: `Price: low to high`,
  PRICE_HL: `Price: high to low`,
  TOPRATED: `Top rated first`
};

export const SORTTYPES = [
  SortType.POPULAR,
  SortType.PRICE_LH,
  SortType.PRICE_HL,
  SortType.TOPRATED
];

export const RatingValues = [5, 4, 3, 2, 1];

export const ErrorType = {
  UNAUTHORIZED: 401,
  BABREQUEST: 400
};

export const TIMEOUT = 5000;
export const URL = `https://htmlacademy-react-3.appspot.com/six-cities`;

export const AuthStatus = {
  NO_AUTH: `NO_AUTH`,
  AUTH: `AUTH`
};

export const LoadingStatus = {
  DEFAULT: `DEFAULT`,
  SUCCESS: `SUCCESS`,
  LOADING: `LOADING`,
  ERROR: `ERROR`
};

export const LoadingStatuses = [
  LoadingStatus.DEFAULT,
  LoadingStatus.SUCCESS,
  LoadingStatus.LOADING,
  LoadingStatus.ERROR
];

export const Url = {
  COMMENTS: `comments`,
  HOTELS: `hotels`,
  LOGIN: `login`,
  NEARBY: `nearby`
};

export const AppRoute = {
  LOGIN: `/login`,
  FAVORITES: `/favorites`,
  OFFER: `/offer`,
  ROOT: `/`,
};
