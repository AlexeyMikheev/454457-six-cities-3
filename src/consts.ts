export enum Feature {
    WIFI= `Wi-Fi`,
    WASHINGMACHINE= `Washing machine`,
    TOWELS= `Towels`,
    HEATING= `Heating`,
    COFFEEMACHINE= `Coffee machine`,
    BABYSET= `Baby seat`,
    KITCHEN= `Kitchen`,
    DISHWASHER= `Dishwasher`,
    CABELTV= `Cabel TV`,
    FRIDGE= `Fridge`
  }

export const FEATURES: Feature[] = [
  Feature.WIFI,
  Feature.WASHINGMACHINE,
  Feature.HEATING,
  Feature.COFFEEMACHINE,
  Feature.BABYSET,
  Feature.DISHWASHER,
  Feature.CABELTV,
  Feature.FRIDGE
];

export enum OfferType {
  APARTMENT= `Apartment`,
  ROOM= `Private room`,
  HOUSE= `House`,
  HOTEL= `Hotel`
}

export const OFFERTYPES: OfferType[] = [
  OfferType.APARTMENT,
  OfferType.ROOM,
  OfferType.HOUSE,
  OfferType.HOTEL
];

export enum ViewMode {
  Main= 0,
  Property= 1,
  Favorite= 2
}

// export const VIEWMODES: ViewMode[] = [
//   ViewMode.Main,
//   ViewMode.Property,
//   ViewMode.Favorite
// ];

export const MIN_RATING = 1;
export const MAX_RATING = 5;

export const MAX_REVIEWS_DISPLAY_COUNT = 10;

export const MAX_IMAGES_DISPLAY_COUNT = 6;
export const MAX_NEAR_DISPLAY_COUNT = 3;

export enum SortType {
  POPULAR= `Popular`,
  PRICE_LH= `Price: low to high`,
  PRICE_HL= `Price: high to low`,
  TOPRATED= `Top rated first`
}

export const SORTTYPES: SortType[] = [
  SortType.POPULAR,
  SortType.PRICE_LH,
  SortType.PRICE_HL,
  SortType.TOPRATED
];

export const RatingValues = [`5`, `4`, `3`, `2`, `1`];

export enum ErrorType {
  UNAUTHORIZED= 401,
  BABREQUEST= 400,
  NOTFOUND= 404
}

export const TIMEOUT = 5000;
export const URL = `https://htmlacademy-react-3.appspot.com/six-cities`;

export enum AuthStatus {
    NO_AUTH= `NO_AUTH`,
    LOADING= `LOADING`,
    AUTH= `AUTH`
  }

export enum LoadingStatus {
  DEFAULT= `DEFAULT`,
  SUCCESS= `SUCCESS`,
  LOADING= `LOADING`,
  ERROR= `ERROR`
}

export const LoadingStatuses: LoadingStatus[] = [
  LoadingStatus.DEFAULT,
  LoadingStatus.SUCCESS,
  LoadingStatus.LOADING,
  LoadingStatus.ERROR
];

export enum Url {
  COMMENTS= `comments`,
  HOTELS= `hotels`,
  FAVORITE= `favorite`,
  LOGIN= `login`,
  NEARBY= `nearby`
}

export enum AppRoute {
  LOGIN= `/login`,
  FAVORITES= `/favorites`,
  OFFER= `/offer`,
  ROOT= `/`,
}


export enum FavoriteState {
  MARKED= 1,
  UNMARKED= 0
}

export enum FavoriteButtonType {
  PROPERTY= 1,
  CARD= 0
}

// export const FavoriteButtonTypes = [
//   FavoriteButtonType.PROPERTY,
//   FavoriteButtonType.CARD
// ];

export enum ImageCardWidth {
  CARD_WIDTH= 260,
  CARD_HEIGHT= 200,
  FAVORITE_WIDTH= 150,
  FAVORITE_HEIGHT= 110
}

export const MIN_COMMENT_LENGTH = 50;
