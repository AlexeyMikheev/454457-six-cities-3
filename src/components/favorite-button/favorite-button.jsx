import React from "react";
import PropTypes from "prop-types";
import {Redirect} from "react-router-dom";
import {FavoriteButtonTypes, FavoriteButtonType, AppRoute} from "../../consts.js";

const FavoriteButton = ({viewType, isMarked, isRedirect, onButtonClick}) => {

  const buttonClass = viewType === FavoriteButtonType.PROPERTY ? `property__bookmark-button` : `place-card__bookmark-button`;
  const iconClass = viewType === FavoriteButtonType.PROPERTY ? `property__bookmark-icon` : `place-card__bookmark-icon`;
  const iconWidth = viewType === FavoriteButtonType.PROPERTY ? 31 : 18;
  const iconHeight = viewType === FavoriteButtonType.PROPERTY ? 33 : 19;

  if (isRedirect) {
    return <Redirect to={AppRoute.LOGIN} />;
  }

  return (
    <button className={`${buttonClass} button ${isMarked ? `${buttonClass}--active` : ``}`} type="button" onClick={onButtonClick}>
      <svg className={iconClass} width={iconWidth} height={iconHeight}>
        <use xlinkHref="#icon-bookmark"></use>
      </svg>
      <span className="visually-hidden">To bookmarks</span>
    </button>
  );
};

FavoriteButton.propTypes = {
  isMarked: PropTypes.bool.isRequired,
  onButtonClick: PropTypes.func.isRequired,
  isRedirect: PropTypes.bool.isRequired,
  viewType: PropTypes.oneOf(FavoriteButtonTypes)
};

export default React.memo(FavoriteButton);
