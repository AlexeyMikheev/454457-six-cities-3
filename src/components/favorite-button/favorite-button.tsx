import * as React from "react";
import {connect} from "react-redux";
import history from "../../history.js";
import {FavoriteButtonType, AppRoute, FavoriteState} from "../../consts";
import {isUserAuthorized} from "../../reducer/user/selectors.js";
import {Operation as DataOperation} from "../../reducer/data/data.js";

interface Props{
  isAuthorized: boolean;
  isMarked: boolean;
  viewType: FavoriteButtonType;
  offerId: number;
  setFavorite: (offerId, state) => void;
}

class FavoriteButton extends React.PureComponent<Props> {
  constructor(props) {
    super(props);

    this.onButtonClick = this.onButtonClick.bind(this);
  }

  onButtonClick(evt) {
    evt.preventDefault();

    const {isAuthorized} = this.props;

    if (!isAuthorized) {
      history.push(AppRoute.LOGIN);
      return;
    }

    const {offerId, isMarked, setFavorite} = this.props;

    const markedValue = isMarked ? FavoriteState.UNMARKED : FavoriteState.MARKED;

    setFavorite(offerId, markedValue);
  }


  render() {
    const {viewType, isMarked} = this.props;

    const buttonClass = viewType === FavoriteButtonType.PROPERTY ? `property__bookmark-button` : `place-card__bookmark-button`;
    const iconClass = viewType === FavoriteButtonType.PROPERTY ? `property__bookmark-icon` : `place-card__bookmark-icon`;
    const iconWidth = viewType === FavoriteButtonType.PROPERTY ? 31 : 18;
    const iconHeight = viewType === FavoriteButtonType.PROPERTY ? 33 : 19;

    return (
      <button className={`${buttonClass} button ${isMarked ? `${buttonClass}--active` : ``}`} type="button" onClick={this.onButtonClick}>
        <svg className={iconClass} width={iconWidth} height={iconHeight}>
          <use xlinkHref="#icon-bookmark"></use>
        </svg>
        <span className="visually-hidden">To bookmarks</span>
      </button>
    );
  }
}

const mapStateToProps = (state) => ({
  isAuthorized: isUserAuthorized(state),
});

const mapDispatchToProps = {
  setFavorite: DataOperation.setFavorite
};

export {FavoriteButton};
export default connect(mapStateToProps, mapDispatchToProps)(FavoriteButton);
