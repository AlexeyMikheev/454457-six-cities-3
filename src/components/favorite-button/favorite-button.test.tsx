import * as React from 'react';
import {mount} from "enzyme";
import {FavoriteButton} from './favorite-button';
import {FavoriteButtonType} from '../../consts';


it(`FavoriteButton snapshot (PROPERTY)`, () => {

  const div = document.createElement(`div`);
  document.body.appendChild(div);

  const favoriteButton = mount(<FavoriteButton
    isAuthorized={true}
    isMarked={true}
    viewType={FavoriteButtonType.PROPERTY}
    offerId={1}
    setFavorite={jest.fn()}
  />, {attachTo: div});

  expect(favoriteButton).toMatchSnapshot();
});

it(`FavoriteButton snapshot (CARD)`, () => {

  const div = document.createElement(`div`);
  document.body.appendChild(div);

  const favoriteButton = mount(<FavoriteButton
    isAuthorized={true}
    isMarked={true}
    viewType={FavoriteButtonType.CARD}
    offerId={1}
    setFavorite={jest.fn()}
  />, {attachTo: div});

  expect(favoriteButton).toMatchSnapshot();
});
