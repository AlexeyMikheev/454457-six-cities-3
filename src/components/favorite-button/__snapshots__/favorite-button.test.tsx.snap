// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteButton snapshot (CARD) 1`] = `
<FavoriteButton
  isAuthorized={true}
  isMarked={true}
  offerId={1}
  setFavorite={[MockFunction]}
  viewType={0}
>
  <button
    className="place-card__bookmark-button button place-card__bookmark-button--active"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="place-card__bookmark-icon"
      height={19}
      width={18}
    >
      <use
        xlinkHref="#icon-bookmark"
      />
    </svg>
    <span
      className="visually-hidden"
    >
      To bookmarks
    </span>
  </button>
</FavoriteButton>
`;

exports[`FavoriteButton snapshot (PROPERTY) 1`] = `
<FavoriteButton
  isAuthorized={true}
  isMarked={true}
  offerId={1}
  setFavorite={[MockFunction]}
  viewType={1}
>
  <button
    className="property__bookmark-button button property__bookmark-button--active"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="property__bookmark-icon"
      height={33}
      width={31}
    >
      <use
        xlinkHref="#icon-bookmark"
      />
    </svg>
    <span
      className="visually-hidden"
    >
      To bookmarks
    </span>
  </button>
</FavoriteButton>
`;
